core/kazoo_amqp/src/api/kapi_im.erl

```diff
@@ -11,137 +11,210 @@
-        'false' -> {'error', "Proplist failed validation for inbound"}
-    end;
-inbound(JObj) -> inbound(kz_json:to_proplist(JObj)).
+-ifdef(TEST).
+-export([routing_key/4
+        ]).
+-endif.
+
+%%------------------------------------------------------------------------------
+%% @doc Get all API definitions of this module.
+%% @end
+%%------------------------------------------------------------------------------
+-spec api_definitions() -> kapi_definition:apis().
+api_definitions() ->
+    [inbound_definition()
+    ,outbound_definition()
+    ].
+
+%%------------------------------------------------------------------------------
+%% @doc Get API definition of the given `Name'.
+%% @see api_definitions/0
+%% @end
+%%------------------------------------------------------------------------------
+-spec api_definition(kz_term:text()) -> kapi_definition:api().
+api_definition(Name) when not is_binary(Name) ->
+    api_definition(kz_term:to_binary(Name));
+api_definition(<<"inbound">>) ->
+    inbound_definition();
+api_definition(<<"outbound">>) ->
+    outbound_definition().
+
+-spec inbound_definition() -> kapi_definition:api().
+inbound_definition() ->
+    EventName = <<"inbound">>,
+    Category = ?EVENT_CATEGORIES,
+    Setters = [{fun kapi_definition:set_name/2, EventName}
+              ,{fun kapi_definition:set_friendly_name/2, <<"Inbound">>}
+              ,{fun kapi_definition:set_description/2, <<"SMS MMS Inbound">>}
+              ,{fun kapi_definition:set_category/2, Category}
+              ,{fun kapi_definition:set_build_fun/2, fun inbound/1}
+              ,{fun kapi_definition:set_validate_fun/2, fun inbound_v/1}
+              ,{fun kapi_definition:set_publish_fun/2, fun publish_inbound/1}
+              ,{fun kapi_definition:set_binding/2, fun publish_inbound_routing_key/1}
+              ,{fun kapi_definition:set_required_headers/2, [<<"Body">>
+                                                            ,<<"From">>
+                                                            ,<<"Message-ID">>
+                                                            ,<<"To">>
+                                                            ]}
+              ,{fun kapi_definition:set_optional_headers/2, [<<"Account-ID">>
+                                                            ,<<"Charges">>
+                                                            ,<<"Custom-SIP-Headers">>
+                                                            ,<<"Custom-Vars">>
+                                                            ,<<"Request">>
+                                                            ,<<"Route-ID">>
+                                                            ,<<"Route-Type">>
+                                                            ,<<"System-ID">>
+                                                            ,<<"Transaction-ID">>
+                                                            ]}
```

Comment:
why are you defining the headers here and not reuse the macros instead ?

Reply:
Because only the definition function should know about those, if we leave them on the macros then they could be used by any other function instead of using `kapi_definition` module getter functions which is the intention with these changes.

Reply:
that doesn't make sense. i prefer to have the macros